src/index.js

```diff
@@ -1,390 +1,586 @@
+}
 
-      for (var i = 0; i < node.children.length; i++) {
-        element.appendChild(
-          createElement(
-            (node.children[i] = resolveNode(node.children[i])),
-            isSvg
-          )
-        )
-      }
+var createVNode = function(name, props, children, element, key, type) {
+  return {
+    name: name,
+    props: props,
+    children: children,
+    element: element,
+    key: key,
+    type: type
+  }
+}
 
-      for (var name in attributes) {
-        updateAttribute(element, name, attributes[name], null, isSvg)
-      }
-    }
+var createTextVNode = function(text, element) {
+  return createVNode(text, EMPTY_OBJECT, EMPTY_ARRAY, element, null, TEXT_NODE)
+}
 
-    return element
-  }
+var recycleChild = function(element) {
+  return element.nodeType === 3 // Node.TEXT_NODE
+    ? createTextVNode(element.nodeValue, element)
+    : recycleElement(element)
+}
 
-  function updateElement(element, oldAttributes, attributes, isSvg) {
-    for (var name in clone(oldAttributes, attributes)) {
-      if (
-        attributes[name] !==
-        (name === "value" || name === "checked"
-          ? element[name]
-          : oldAttributes[name])
-      ) {
-        updateAttribute(
-          element,
-          name,
-          attributes[name],
-          oldAttributes[name],
-          isSvg
-        )
-      }
-    }
+var recycleElement = function(element) {
+  return createVNode(
+    element.nodeName.toLowerCase(),
+    EMPTY_OBJECT,
+    map.call(element.childNodes, recycleChild),
+    element,
+    null,
+    RECYCLED_NODE
+  )
+}
 
-    var cb = isRecycling ? attributes.oncreate : attributes.onupdate
-    if (cb) {
-      lifecycle.push(function() {
-        cb(element, oldAttributes)
-      })
-    }
-  }
+var patch = function(container, element, lastNode, nextNode, eventProxy) {
+  var lifecycle = []
 
-  function removeChildren(element, node) {
-    var attributes = node.attributes
-    if (attributes) {
-      for (var i = 0; i < node.children.length; i++) {
-        removeChildren(element.childNodes[i], node.children[i])
-      }
+  patchElement(container, element, lastNode, nextNode, lifecycle, eventProxy)
 
-      if (attributes.ondestroy) {
-        attributes.ondestroy(element)
-      }
-    }
-    return element
-  }
+  while (lifecycle.length > 0) lifecycle.pop()()
 
-  function removeElement(parent, element, node) {
-    function done() {
-      parent.removeChild(removeChildren(element, node))
-    }
+  return nextNode
+}
 
-    var cb = node.attributes && node.attributes.onremove
-    if (cb) {
-      cb(element, done)
-    } else {
-      done()
+export var h = function(name, props) {
+  var node
+  var rest = []
+  var children = []
+  var length = arguments.length
+
+  while (length-- > 2) rest.push(arguments[length])
+
+  if ((props = props == null ? {} : props).children != null) {
+    if (rest.length <= 0) {
+      rest.push(props.children)
     }
+    delete props.children
```

Comment:
A tip from TS core about `delete`:

```js
// Using 'delete' on an object causes V8 to put the object in dictionary mode.
// This disables creation of hidden classes, which are expensive when an object is
// constantly changing shape.
map.__ = undefined;
delete map.__;
```

Source: [TS core](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/core.ts#L73-L78)
Perf: [delete vs undefined vs null](https://jsperf.com/delete-vs-undefined-vs-null/16)

Thoughts?

Reply:
Definitely `delete` is what we want _if_ we want ({ children, ...props }). If we decide to go with (props, children), then sure.

Reply:
Sure,but in that case, does it make sense to you to set undefined only?